### ISSUE TITLE
Feature: multiple balanced assets

### ISSUE THREAD

**Issue by kwannoel** [kwannoel/urban-octo-carnival#7]
In GitLab by @fahree on Mar 11, 2020, 14:52

Add support for interactions that involve multiple assets, wherein the fact that assets are balanced at all points is guaranteed by the language. Some assets may be fungible, others not.

Stretch goal: have a common abstraction for goods and services, so that transfer of goods is a service, and services can be sold.

In `gerbil-ethereum` `assets.ss`, a method for setting up the "expected-to-be-deposited-so-far" variable, updating that variable on deposit, and committing that on a transaction? And a similar thing for setting up withdrawls to each of the participants, and committing those at the end of a transaction

Subtasks:
 - [x] Move block-context back to glow
 - [x] Multiple assets in block context
   * [x] Tracking balances of multiple assets in participant-runtime
   * [x] Tracking balances to "variables" in the evm-runtime
 - [x] Asset type with methods for committing transfers at the end of a transaction
 - [x] Add assets to InteractionAgreement
 - [x] Use ethereum-config network to filter the asset options in CLI
 - [x] Pass assets through CLI into participant-runtime and consensus-code-generation
 - [x] In runtime, set up asset environment (including block contexts)
   * [x] In runtime, set up asset environment
   * [x] including block contexts
 - [x] For deposit/withdraw, use asset-environment/block-contexts to track balances
 - [x] On participant committing transactions, check balances are valid
   * [x] Check deposit native balance for native assets
   * [x] Check deposit erc20 balances for erc20 assets
   * [x] Check withdraw contract balances (balance variable, not necessarily contract balance, for multiple interactions per contract)
 - [x] On participant committing transactions, transfers
   * [x] Attach CALLVALUE for native assets
   * [x] Check for zero, approve transfer-from for erc20 assets
 - [x] On consensus committing transactions, use asset-environment to deposit/withdraw the asset amounts for the transaction
   * [x] Check CALLVALUE for native asset deposits
   * [x] Check for zero, invoke transfer-from for erc20 asset deposits
   * [x] Check for zero, invoke transfer for native asset withdraws
   * [x] Check for zero, invoke transfer for erc20 asset withdraws
 - [x] Test all of this

Related branches:
 - `gerbil-ethereum`: https://github.com/AlexKnauth/gerbil-ethereum/tree/assets4
 - `glow`: https://gitlab.com/mukn/glow/-/commits/assets5

**Comment by kwannoel** [comment]
In GitLab by @fahree on Mar 11, 2020, 14:54

changed the description

**Comment by kwannoel** [comment]
In GitLab by @fahree on Nov 7, 2020, 03:42

assigned to @fahree

**Comment by kwannoel** [comment]
In GitLab by @ApolloUnicorn on Feb 13, 2021, 01:35

marked this issue as related to #115

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Mar 2, 2021, 04:04

changed the description

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Mar 2, 2021, 04:05

changed the description

**Comment by kwannoel** [comment]
In GitLab by @ApolloUnicorn on Mar 4, 2021, 00:15

assigned to @AlexKnauth

**Comment by kwannoel** [comment]
In GitLab by @ApolloUnicorn on Mar 4, 2021, 00:15

removed the relation with #115

**Comment by kwannoel** [comment]
In GitLab by @ApolloUnicorn on Mar 4, 2021, 00:15

marked this issue as related to #115

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Mar 9, 2021, 03:31

marked this issue as related to #106

**Comment by kwannoel** [comment]
In GitLab by @ApolloUnicorn on Mar 17, 2021, 23:37

unassigned @fahree

**Comment by kwannoel** [comment]
In GitLab by @ApolloUnicorn on Mar 17, 2021, 23:41

removed the relation with #106

**Comment by kwannoel** [comment]
In GitLab by @ApolloUnicorn on Mar 26, 2021, 23:28

@AlexKnauth would you please add subtasks to this ticket that outline what is left? I am having a hard time understand the steps/progress.

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Mar 31, 2021, 23:49

changed the description

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 6, 2021, 02:07

changed the description

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 6, 2021, 02:08

changed the description

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 7, 2021, 23:39

marked the task **Add assets to InteractionAgreement** as completed

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 7, 2021, 23:45

changed the description

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 20, 2021, 02:34

marked the task **Pass assets through CLI into participant-runtime and consensus-code-generation** as completed

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 21, 2021, 23:37

marked the task **Asset type with methods for committing transfers at the end of a transaction** as completed

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 27, 2021, 02:07

changed the description

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 28, 2021, 23:21

marked the task **including block contexts** as completed

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 28, 2021, 23:21

marked the task **In runtime, set up asset environment (including block contexts)** as completed

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Apr 28, 2021, 23:21

marked the task **For deposit/withdraw, use asset-environment/block-contexts to track balances** as completed

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on May 6, 2021, 01:37

changed the description

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on May 8, 2021, 04:53

marked the task **Use ethereum-config network to filter the asset options in CLI** as completed

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on May 19, 2021, 01:26

changed the description

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on May 19, 2021, 23:23

changed the description

**Comment by kwannoel** [comment]
In GitLab by @ApolloUnicorn on May 20, 2021, 18:59

@AlexKnauth are there any of the checklist items that are actually complete? I know there was some pairing last week. I am trying to identify if we are blocked on any particular piece of the plan specified.

**Comment by kwannoel** [comment]
In GitLab by @ApolloUnicorn on May 20, 2021, 19:01

@fahree I believe this was integral to the May delivery, but it's not clear what this blocks/is blocked by. Is there a subset of the open subtasks above that could be segmented out for post-May deadlines?

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Jun 15, 2021, 23:08

marked the task **Tracking balances to "variables" in the evm-runtime** as completed

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Jun 15, 2021, 23:08

marked the task **Multiple assets in block context** as completed

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Jun 15, 2021, 23:14

changed the description

**Comment by kwannoel** [comment]
In GitLab by @AlexKnauth on Aug 4, 2021, 21:59

changed the description

**Comment by kwannoel** [comment]
In GitLab by @alex413 on Aug 17, 2021, 06:11

assigned to @isd

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:46

marked the task **Check withdraw contract balances (balance variable, not necessarily contract balance, for multiple interactions per contract)** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:46

marked the task **Check deposit native balance for native assets** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:46

marked the task **Check deposit erc20 balances for erc20 assets** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:46

marked the task **On participant committing transactions, check balances are valid** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:46

marked the task **Check for zero, approve transfer-from for erc20 assets** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:46

marked the task **On participant committing transactions, transfers** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:47

marked the task **Check for zero, invoke transfer-from for erc20 asset deposits** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:47

marked the task **Check for zero, invoke transfer for native asset withdraws** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:47

marked the task **Check for zero, invoke transfer for erc20 asset withdraws** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:47

marked the task **On consensus committing transactions, use asset-environment to deposit/withdraw the asset amounts for the transaction** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:47

marked the task **Test all of this** as completed

**Comment by kwannoel** [comment]
In GitLab by @isd on Aug 25, 2021, 06:51

mentioned in merge request !168
